File: Python/ByTier/C/2006_C_Eri_and_Expanded_Sets.py
# Problem: CF 2006 C - Eri and Expanded Sets
# https://codeforces.com/contest/2006/problem/C

"""
Algorithm: 
- For each subarray, we check if it can be expanded to form a consecutive set.
- The key insight is that for a set to be expandable into a consecutive set, 
  the GCD of differences between adjacent elements must be 1 or a power of 2.
- We use a sliding window approach with a data structure to track the GCDs of 
  differences and update the count of valid subarrays accordingly.
- Time Complexity: O(n log M) where n is the length of the array and M is the maximum element.
- Space Complexity: O(n) for storing intermediate results and differences.

Techniques:
- Sliding window with GCD computation
- Bit manipulation to check if a number is a power of 2
"""

import math
import sys


def solve() -> None:
    it = iter(sys.stdin.buffer.read().split())
    t = int(next(it))
    out = []
    for _ in range(t):
        n = int(next(it))
        a = [int(next(it)) for _ in range(n)]
        ans = n
        if n == 1:
            out.append(str(ans))
            continue
        # Compute difference array
        d = [abs(a[i + 1] - a[i]) for i in range(n - 1)]
        # cur keeps track of (gcd, count) pairs for current window
        cur = []
        for val in d:
            # Start new list of (gcd, count) pairs
            nxt = [(val, 1)]
            # Process previous pairs and update with new gcd
            for g, cnt in cur:
                ng = math.gcd(g, val)
                if nxt[-1][0] == ng:
                    # If same gcd, increment count
                    nxt[-1] = (ng, nxt[-1][1] + cnt)
                else:
                    # Otherwise add new gcd
                    nxt.append((ng, cnt))
            # Check if current pair (gcd, count) leads to a valid consecutive set
            for g, cnt in nxt:
                if g == 0 or (g & (g - 1)) == 0:
                    # If g is 0 or a power of two, we add count to result
                    ans += cnt
            # Update current list
            cur = nxt
        out.append(str(ans))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    solve()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/C/2006_C_Eri_and_Expanded_Sets.cpp
// Problem: CF 2006 C - Eri and Expanded Sets
// https://codeforces.com/contest/2006/problem/C

/*
  Problem: C. Eri and Expanded Sets
  Algorithms/Techniques: Sliding window with GCD, bit manipulation for powers of two

  Time Complexity: O(N * log(max(a[i])) + N) per test case
  Space Complexity: O(N) for arrays and structures

  The task is to count the number of subarrays such that after expanding the set using the given operation,
  the final set can become consecutive. The key insight is that the expansion is possible if and only if
  the GCD of differences between adjacent elements in the sorted subarray is 1 or a power of 2.
  We use a sliding window technique to maintain counts of distinct GCDs and check conditions.
*/

#include <stdint.h>
#include <stdlib.h>
#include <algorithm>
#include <iostream>

using namespace std;
typedef int64_t i64;
const int N = 4e5 + 5, B = 45;
int t, n, a[N], b[N], cn;
i64 ans;
struct node {
  int v, c;
} c[B];
void solve() {
  cin >> n;
  ans = n;  // Every single element subarray is brilliant
  cn = 0;
  for (int i = 1; i <= n; i++)
    cin >> a[i], b[i] = abs(a[i] - a[i - 1]);  // Compute differences between adjacent elements
  for (int i = 2; i <= n; i++) {
    for (int j = 1; j <= cn; j++)
      c[j].v = __gcd(c[j].v, b[i]);  // Update GCDs with current difference
    c[++cn] = {b[i], 1};  // Add the new difference to our structure
    int m = cn;
    cn = 1;
    for (int j = 2; j <= m; j++) {
      if (c[j].v == c[cn].v)
        c[cn].c += c[j].c;  // Merge same GCDs
      else
        c[++cn] = c[j];
    }
    for (int j = 1; j <= cn; j++)
      if (c[j].v == 0 || (c[j].v & (c[j].v - 1)) == 0)  // Check for zero or power of two
        ans += c[j].c;  // Increment answer if condition satisfied
  }
  cout << ans << '\n';
}
int main() {
  ios ::sync_with_stdio(false);
  cin.tie(nullptr), cout.tie(nullptr);
  cin >> t;
  while (t--)
    solve();
  return 0;
}


// https://github.com/VaHiX/codeForces/